### PR TITLE
Show author changes in exercise objects

### PR DESCRIPTION
**Pull request by ImTheTom**
# Proposed Changes

- Create new abstract model called AbstractHistoryMixin. Used to represent extractions from historical records.
- Adds a command to add to change author history for exercises and exercises bases

## Please check that the PR fulfills these requirements

- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Added yourself to AUTHORS.rst

### Other questions

* Placing it up for review early to see if I'm on the right track

### Showcase
Exercise object no changes
```
{
    "id": 1,
    "uuid": "b83e3d85-a53d-4939-a61c-7baa2e94d358",
    "name": "Trizeps Seildrücken",
    "exercise_base": 659,
    "description": "<p>Stelle dich mit schulterbreitem Stand vor dem Kabelzug, beuge den Rücken sehr leicht nach vorn und greife das Seil. Ziehe nun nach unten, wobei die Ellenbogen sich während der ganzen Übung nicht von der Stelle bewegen sollen. Wenn die Arme ausgestreckt sind, drehe die Hände nach außen. Gehe langsam zur Ausgansposition zurück.</p>",
    "creation_date": "2013-05-05",
    "category": 8,
    "muscles": [
        5
    ],
    "muscles_secondary": [],
    "equipment": [],
    "language": 1,
    "license": 1,
    "license_author": "James",
    "variations": [
        1,
        63
    ],
    "author_history": []
}
```
Add to license history table
```
python manage.py change-author --author-name Tom --exercise-id 1
```
Exercise now shows
```
{
    "id": 1,
    "uuid": "b83e3d85-a53d-4939-a61c-7baa2e94d358",
    "name": "Trizeps Seildrücken",
    "exercise_base": 659,
    "description": "<p>Stelle dich mit schulterbreitem Stand vor dem Kabelzug, beuge den Rücken sehr leicht nach vorn und greife das Seil. Ziehe nun nach unten, wobei die Ellenbogen sich während der ganzen Übung nicht von der Stelle bewegen sollen. Wenn die Arme ausgestreckt sind, drehe die Hände nach außen. Gehe langsam zur Ausgansposition zurück.</p>",
    "creation_date": "2013-05-05",
    "category": 8,
    "muscles": [
        5
    ],
    "muscles_secondary": [],
    "equipment": [],
    "language": 1,
    "license": 1,
    "license_author": "James",
    "variations": [
        1,
        63
    ],
    "author_history": [
        "Tom"
    ]
}
```
python manage.py change-author --author-name Tom --exercise-id 1
```
{
    "id": 1,
    "uuid": "b83e3d85-a53d-4939-a61c-7baa2e94d358",
    "name": "Trizeps Seildrücken",
    "exercise_base": 659,
    "description": "<p>Stelle dich mit schulterbreitem Stand vor dem Kabelzug, beuge den Rücken sehr leicht nach vorn und greife das Seil. Ziehe nun nach unten, wobei die Ellenbogen sich während der ganzen Übung nicht von der Stelle bewegen sollen. Wenn die Arme ausgestreckt sind, drehe die Hände nach außen. Gehe langsam zur Ausgansposition zurück.</p>",
    "creation_date": "2013-05-05",
    "category": 8,
    "muscles": [
        5
    ],
    "muscles_secondary": [],
    "equipment": [],
    "language": 1,
    "license": 1,
    "license_author": "James",
    "variations": [
        1,
        63
    ],
    "author_history": [
        "Tom"
    ]
}
```
python manage.py change-author --author-name James --exercise-id 1
```
{
    "id": 1,
    "uuid": "b83e3d85-a53d-4939-a61c-7baa2e94d358",
    "name": "Trizeps Seildrücken",
    "exercise_base": 659,
    "description": "<p>Stelle dich mit schulterbreitem Stand vor dem Kabelzug, beuge den Rücken sehr leicht nach vorn und greife das Seil. Ziehe nun nach unten, wobei die Ellenbogen sich während der ganzen Übung nicht von der Stelle bewegen sollen. Wenn die Arme ausgestreckt sind, drehe die Hände nach außen. Gehe langsam zur Ausgansposition zurück.</p>",
    "creation_date": "2013-05-05",
    "category": 8,
    "muscles": [
        5
    ],
    "muscles_secondary": [],
    "equipment": [],
    "language": 1,
    "license": 1,
    "license_author": "James",
    "variations": [
        1,
        63
    ],
    "author_history": [
        "James",
        "Tom"
    ]
}
```

### CODE REVIEW

**Comment by rolandgeider** [comment]
Looks good! This was easier than what I had thought, I will try to take a closer look later today or so.

I'm not sure whether we need to expose the complete history: if you change the description twice your name will appear twice and while we need to keep this data internally, it's not so useful for consumers over the api.

**Comment by ImTheTom** [comment]
Yeah I can see the advantage of having just the distinct authors.

Another improvment to the MR would be to have these authors on all exercise API endpoints that it would make sense on. I guess that would be any already with the original license author on it.

**Comment by rolandgeider** [comment]
I'm also thinking we could move `license_author_history` to a class, then we could easily expose the authors for all models that have a history

**Comment by ImTheTom** [comment]
Yeah sure. Will be easier to maintain. I'll close this off and go with that approach.

**Comment by rolandgeider** [comment]
You don't need to close this or, Just push your changes

**Comment by ImTheTom** [comment]
Hey, would this be more suitable.

**Comment by rolandgeider** [comment]
wouldn't the new LicenseAuthorHistory more or less duplicate some of the logic that the history plugin already does? What I meant with a class was to write a simple mixin that we could then just plug in in all the models that have a history.

```python
class HistoryMixin():
    @property
    def author_history(self):
        ...
        return {}
        
        
class Exercise(AbstractLicenseModel, models.Model, HistoryMixin):
    ....
```

**Comment by rolandgeider** [comment]
I should also finally fix the tests in the crowdsourcing branch, it's really annoying having constantly red checks!

**Comment by ImTheTom** [comment]
Oh yeah I'll look at that approach :+1:.

**Comment by ImTheTom** [comment]
I think it's ready again for an initial review before proceeding :).

File: wger/utils/models.py [hunk 1]
```diff
@@ -86,3 +86,22 @@ class Meta:
...
+    Utilized in conjunction with simple_history's HistoricalRecords.
+    """
+
+    class Meta:
+        abstract = True
+
+    @property
+    def author_history(self):
+        out = []
+        for history in self.history.all():
```
[the remark this answers is not in the document]
Comment (ImTheTom) [reply wger/utils/models.py:104]: Yeah sure. I'll continue expanding out using this as a base.

**Comment by ImTheTom** [comment]
Hey man,

Is there anything else you would like from this change?

**Comment by rolandgeider** [comment]
Hi! I just haven't found any time, but this looks ready :)

**Comment by ImTheTom** [comment]
Yeah all good :)

**Pull request merged**